server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java

```diff
@@ -59,68 +54,22 @@ protected PlainListenableActionFuture() {}
 
     @Override
     public void addListener(final ActionListener<T> listener) {
-        internalAddListener(listener);
-    }
-
-    @Override
-    protected void done() {
-        super.done();
-        synchronized (this) {
-            executedListeners = true;
-        }
-        Object listeners = this.listeners;
-        if (listeners != null) {
-            if (listeners instanceof List) {
-                List list = (List) listeners;
-                for (Object listener : list) {
-                    executeListener((ActionListener<T>) listener);
-                }
+        whenComplete((val, throwable) -> {
+            if (throwable == null) {
+                listener.onResponse(val);
```

Comment:
old code caught exceptions here and routed them to listener.onFailure()

Reply:
Good catch. I'm not a fan of having this anti-pattern in library code but I'm still making this change for BWC purposes. In the future (pun intended) we can also think about adding these methods directly to BaseFuture as every Future has now listening capabilities.